# E6. Complementary slackness

## Primal LP Problem : 

\begin{align*}
\text{Maximize} \quad & x_{1} + 4x_{2} + 2x_{3} \\
\text{subject to} \quad & 5x_{1} + 2x_{2} + 2x_{3} \leq 145, \\
& 4x_{1} + 8x_{2} - 8x_{3} \leq 260, \\
& x_{1} + x_{2} + 4x_{3} \leq 190, \\
& x_{1}, x_{2}, x_{3} \geq 0.
\end{align*}


## Dual LP Problem :

\begin{align*}
\text{Minimize} \quad & 145 y_{1} + 260 y_{2} + 190 y_{3} \\
\text{subject to} \quad & 5 y_{1} + 4 y_{2} + y_{3} \geq 1, \\
& 2 y_{1} + 8 y_{2} + y_{3} \geq 4, \\
& 2 y_{1} - 8 y_{2} + 4 y_{3} \geq 2, \\
& y_1, y_2, y_3 \geq 0
\end{align*}

## Standard form :

\begin{align*}
\text{Maximize} \quad & z -145y_1 - 260y_2 - 190y_3 = 0 \\
\text{subject to} \quad & -5y_1 - 4y_2 - y_3 \leq -1 \\
& -2y_1 - 8y_2 - y_3 \leq -4 \\
& -2y_1 + 8y_2 - 4y_3 \leq -2 \\
& y \geq 0
\end{align*}

## Verify that  $Q=(x_1,x_2,x_3)=(0,52.5,20)$ is a feasible solution for the primal

In [2]:
x1, x2, x3 = 0, 52.5, 20

# Constraints of the primal problem
constraint1 = 5*x1 + 2*x2 + 2*x3
constraint2 = 4*x1 + 8*x2 - 8*x3
constraint3 = x1 + x2 + 4*x3

# Values for constraints based on Q
constraint1_value = constraint1 <= 145
constraint2_value = constraint2 <= 260
constraint3_value = constraint3 <= 190
nonnegative_values = x1 >= 0 and x2 >= 0 and x3 >= 0

# Check if all constraints are satisfied
feasible_solution = all([constraint1_value, constraint2_value, constraint3_value, nonnegative_values])

feasible_solution, constraint1, constraint2, constraint3

(True, 145.0, 260.0, 132.5)

## Use CS to determine a candidate solution to the dual

In [7]:
import sympy as sp
# Define the symbols for y1 and y2
y1, y2 = sp.symbols('y1 y2', nonnegative=True)

# Equations derived from Complementary Slackness conditions
eq1 = sp.Eq(2*y1 + 8*y2, 4)
eq2 = sp.Eq(2*y1 - 8*y2, 2)

# Solve the system of equations
dual_solution = sp.solve((eq1, eq2), (y1, y2))
dual_solution

{y1: 3/2, y2: 1/8}

The third primal constraint is slack because the calculated left-hand side is less than the right-hand side, 
132.5<190. This means that the corresponding dual variable y3 must be zero.
That means we have $(y_1,y_2,y_3)=(3/2,1/8,0)$

## Is $Q$ the solution for the primal problem?

To determine if \( $Q$ = (x_1, x_2, x_3) = (0, 52.5, 20) \) is an optimal solution using the Simplex method, we can set up the Simplex tableau and solve it, watching for the signs of the reduced costs in the final tableau.

### Simplex Tableau Setup

We need to set up the initial tableau for the primal problem. Given the constraints:
- \( 5x_1 + 2x_2 + 2x_3 < 145 \)
- \( 4x_1 + 8x_2 - 8x_3 < 260 \)
- \( x_1 + x_2 + 4x_3 < 190 \)

We introduce slack variables \( s_1, s_2, s_3 \) for each constraint, respectively. The objective function is:
- \( z = x_1 + 4x_2 + 2x_3 \)

### Objective and Constraints Transformation
Transforming the constraints and objective function:

- Objective: Maximize \( z = x_1 + 4x_2 + 2x_3 \)
- Constraint 1: \( 5x_1 + 2x_2 + 2x_3 + s_1 = 145 \)
- Constraint 2: \( 4x_1 + 8x_2 - 8x_3 + s_2 = 260 \)
- Constraint 3: \( x_1 + x_2 + 4x_3 + s_3 = 190 \)

For an initial basis using slack variables \( s_1, s_2, s_3 \), the tableau can be initialized. The Simplex tableau includes these rows plus a row for the objective function, where non-basis variables have a cost of zero in the basis.

Let's construct this initial tableau and perform iterations to see if $Q$ is optimal. For this demonstration, we'll set up the tableau and execute a few steps to check the optimality.

After running the Simplex method on the given tableau, we have reached an optimal solution, as evidenced by the absence of negative coefficients in the objective function row (the last row) for the decision variables (x1, x2, x3). This means that the solution is optimal according to the Simplex criteria. The tableau has been transformed such that:

- **x1** corresponds to a value of **20**, pivoted and part of the solution (first column is a basis column).
- **x2** corresponds to a value of **52.5**, also a basis column.
- **s3** (slack variable for the third constraint) corresponds to **57.5**.

The final objective value is **250**, as seen in the last element of the objective function row. This confirms that \( $Q$ = (x_1, x_2, x_3) = (0, 52.5, 20) \) not only is a feasible solution, as we saw before, but it is also the optimal solution for the primal problem, based on the Simplex calculations showing no further direction for improvement (no negative coefficients in the decision variable columns of the objective function row). 

Thus, $Q$ indeed maximizes the objective function under the given constraints.